NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/mcxassemble.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "mcxassemble" 1 "16 May 2014" "mcxassemble 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
mcxassemble \- transform raw cooccurrence data to mcl matrix format\&.
.SH SYNOPSIS

\fBmcxassemble\fP
\fB-b\fP base (\fIbase name\fP)
\fB[-o\fP fname (\fIwrite to file fname\fP)\fB]\fP
\fB[--write-binary\fP (\fIwrite output in binary format\fP)\fB]\fP
\fB[--map\fP (\fIapply base\&.map\fP)\fB]\fP
\fB[-raw-tf\fP (\fIapply transform spec to input\fP)\fB]\fP
\fB[-rv\fP MODE (\fIrepeated vectors\fP)\fB]\fP
\fB[-re\fP MODE (\fIrepeated entries\fP)\fB]\fP
\fB[-ri\fP MODE (\fIadding mirror image\fP)\fB]\fP
\fB[-r\fP MODE (\fIrepeated entries/vectors/images\fP)\fB]\fP
\fB[-prm-tf\fP (\fIapply transform spec to primary matrix\fP)\fB]\fP
\fB[-sym-tf\fP (\fIapply transform spec to symmetrified matrix\fP)\fB]\fP
\fB[-q\fP (\fIquiet mode\fP)\fB]\fP

The options above embody the default setup when using mcxassemble\&.
There are many more options which mostly provide subtly different
ways of doing input/output, set warning levels, or regulate
how repeated entries and vectors should be treated\&.
The full list of options is shown below\&.
Read \fBDESCRIPTION\fP for learning about mcxassemble input/output
and the functionality it provides\&.

\fBNOTE\fP
.br
As of release 05-314 \fBmcl(1)\fP is able to cluster label-type input
on the fly\&. In most cases, this will be sufficient\&. Alternatively,
\fBmcxload(1)\fP can be used to map label-type input onto mcl
matrices\&. Consequently, there are likely fewer scenarios nowadays
where \fBmcxassemble\fP is the best solution\&. Consider first whether
\fBmcl\fP in label mode or \fBmcxload\fP can do the job as well\&.

\fBmcxassemble\fP
\fB[-b\fP base (\fIbase name\fP)\fB]\fP
\fB[-hdr\fP fname (\fIread header file\fP)\fB]\fP
\fB[-raw\fP fname (\fIread raw file\fP)\fB]\fP
\fB[--map\fP (\fIapply base\&.map\fP)\fB]\fP
\fB[--cmap\fP (\fIapply base\&.cmap\fP)\fB]\fP
\fB[--rmap\fP (\fIapply base\&.rmap\fP)\fB]\fP
\fB[-map\fP fname (\fIapply fname\fP)\fB]\fP
\fB[-rmap\fP fname (\fIapply fname\fP)\fB]\fP
\fB[-cmap\fP fname (\fIapply fname\fP)\fB]\fP
\fB[-tag\fP tag (\fIapply base\&.tag\fP)\fB]\fP
\fB[-rtag\fP tag (\fIapply base\&.tag\fP)\fB]\fP
\fB[-ctag\fP tag (\fIapply base\&.tag\fP)\fB]\fP
\fB[-skw\fP fname (\fIwrite skew matrix\fP)\fB]\fP
\fB[-prm\fP fname (\fIwrite primary result matrix\fP)\fB]\fP
\fB[--skw\fP (\fIwrite base\&.skw\fP)\fB]\fP
\fB[--prm\fP (\fIwrite base\&.prm\fP)\fB]\fP
\fB[-xo\fP suf (\fIwrite base\&.suf\fP)\fB]\fP
\fB[-o\fP fname (\fIwrite to file fname\fP)\fB]\fP
\fB[-n\fP (\fIdo not write default symmetrized result\fP)\fB]\fP
\fB[-i\fP (\fIread from single data file\fP)\fB]\fP
\fB[-digits\fP int (\fIdigits width\fP)\fB]\fP
\fB[-s\fP (\fIcheck for symmetry\fP)\fB]\fP
\fB[-raw-tf\fP (\fIapply transform spec to input\fP)\fB]\fP
\fB[-rv\fP <mode> (\fIaction for repeated vectors\fP)\fB]\fP
\fB[-re\fP <mode> (\fIaction for repeated entries\fP)\fB]\fP
\fB[-ri\fP <mode> (\fIadding mirror image\fP)\fB]\fP
\fB[-r\fP <mode> (\fIsame for entries and vectors\fP)\fB]\fP
\fB[-prm-tf\fP (\fIapply transform spec to primary matrix\fP)\fB]\fP
\fB[-sym-tf\fP (\fIapply transform spec to symmetrified matrix\fP)\fB]\fP
\fB[--quiet-re\fP (\fIquiet for repeated entries\fP)\fB]\fP
\fB[--quiet-rv\fP (\fIquiet for repeated vectors\fP)\fB]\fP
\fB[-q\fP (\fIthe two above combined\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
.SH DESCRIPTION

\fBmcxassemble\fP enables easy matrix creation from an intermediate raw matrix
format that can easily be constructed from a one-pass-parse of cooccurrence
data\&. The basic setup is as follows\&.

.ZJ 1m 1m "\(bu"
Parse cooccurrence data from some external format\&.
.in -2m
.ZJ 1m 1m "\(bu"
Transform cooccurrence data to raw mcl data as you parse\&.
.in -2m
.ZJ 1m 1m "\(bu"
When done, write out required header and domain information
to a separate file\&. The domain information can be built during
the parsing stage\&.
.in -2m
.ZJ 1m 1m "\(bu"
Use mcxassemble to construct a valid matrix from the raw data
and the header information\&.
.in -2m
.ZJ 1m 1m "\(bu"
Nodes can optionally be relabeled by writing a separate map file to be read
by \fBmcxassemble\fP, which takes the form of a very thin matrix file\&.
.in -2m

The easiest thing to do is to group all input/output files under the same
base name, say\ \&\fBbase\fP\&. A standard way of proceeding, which will lead to
a concise \fBmcxassemble\fP command line, is by creating the input files
\fBbase\&.raw\fP and \fBbase\&.hdr\fP, and optionally the file \fBbase\&.map\fP\&. The
default behaviour of mcxassemble is then to create \fBbase\&.sym\fP as the
resulting matrix file, containing the symmetrized matrix constructed from
the raw input\&.

\fBExample\fP
.br
Suppose \fCblastresult\fP is a file containing blast results\&.
The following two commands construct an mcl matrix file from the file\&.

.di ZV
.in 0
.nf \fC
   mcxdeblast --score=e --sort=a blastresult
   mcxassemble -b blastresult -r max --map
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

\fBmcxdeblast\fP will generate the
files \fCblastresult\&.hdr\fP, \fCblastresult\&.raw\fP, and \fCblastresult\&.map\fP\&.
The \fB--sort=a\fP option will create a map file corresponding
with alphabetic ordering\&. These files are processed by \fBmcxassemble\fP
and it will generate the file \fCblastresult\&.sym\fP\&. The \fB-r\fP
option tells \fBmcxassemble\fP that repeated entries should be maxed;
each time the largest entry seen thus far will be taken\&.

\fBHeader file\fP
.br
This file contains a header as usually found in generic mcl matrix files,
i\&.e\&. the required \fIheader\fP part, and optionally the \fIdomain\fP part(s)
if not all domains are canonical\&. Refer to \fBmcxio(5)\fP for more information\&.
The domain information in the header file will be used to pre-construct a
skeleton matrix and to validate the entries in the raw data file as they
fill the skeleton matrix\&.

\fBRaw input format\fP
.br
The file from which raw input is read should have the raw format as
described in \fBmcxio(5)\fP\&. Simply put; no header specification, no domain
specification, and no matrix introduction syntax is used\&. The file just
contains a listing of vectors\&. An example fragment is the following:

.nf \fC
2  4:0\&.34 1:2\&.8838 4:2\&.328 1:4\&.238 1:12 $
1  2:7\&.8 $
2  1:0\&.01 4:20\&.3 3:2 $
.fi \fR

The listing of vectors need not be sorted, and neither does
a vector itself need to be sorted - the mcl generic matrix format
is actually not different in this respect\&.
Furthermore, duplicate entries and duplicate vectors are allowed\&.
This is in fact again allowed in the generic format, except
that where applications expect generic format warnings will be issued and
duplicate entries will be disregarded\&. \fBmcxassemble\fP allows customizable
behaviour dictating how to merge repeated entries\&.
Refer to the \fB-re\fP,\ \&\fB-rv\fP,\ \&\fB-r\fP
options below\&.

The vectors read by \fBmcxassemble\fP do have to match the domains specified in
the header file\&. The leading index that specifies the column index has to be
present in the column domain; all subsequent indices that specify column
entries have to be present in the row domain\&.

\fIIf one concatenates the contents of the header file and the data file\fP,
the result is \fIalmost but not quite\fP a file containing a matrix in
syntactically correct mcl generic matrix format\&. The parts missing
are the \fC(mclmatrix\fP introduction token, (followed by) the
\fCbegin\fP token, and the closing \fC)\fP token\&.

\fBMap file\fP
.br
This file must contain a map matrix, which is a matrix with the
following properties:

.ZJ 1m 1m "\(bu"
The column domain and row domain are of the same cardinality\&.
.in -2m
.ZJ 1m 1m "\(bu"
Each column has exactly one entry\&.
.in -2m
.ZJ 1m 1m "\(bu"

NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shmx/mcxassemble.c
/*   (C) Copyright 2001, 2002, 2003, 2004, 2005 Stijn van Dongen
 *   (C) Copyright 2006, 2007, 2008, 2009, 2010, 2011 Stijn van Dongen
 *   (C) Copyright 2012, 2013 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/

/* todo; couple suf with xf_sym */


#include <string.h>
#include <stdio.h>

#include "impala/matrix.h"
#include "impala/vector.h"
#include "impala/pval.h"
#include "impala/io.h"
#include "impala/iface.h"
#include "impala/app.h"

#include "mcl/interpret.h"
#include "mcl/transform.h"

#include "util/io.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/alloc.h"
#include "util/types.h"

const char* me = "mcxassemble";

const char* syntax
=  "Usage: mcxassemble [options]\n"
   "Default output is the symmetrized result"
   " of the matrix built from the raw data"
;

enum
{  MY_OPT_BASE
,  MY_OPT_HDR
,  MY_OPT_RAW
,  MY_OPT_SINGLE
,  MY_OPT_OUTPUT
,  MY_OPT_DIGITS
,  MY_OPT_XO
,  MY_OPT_WB
,  MY_OPT_PLUS
,                 MY_OPT_NO
,  MY_OPT_MAP =   MY_OPT_NO  + 2
,  MY_OPT_MAP_
,  MY_OPT_CMAP
,  MY_OPT_CMAP_
,  MY_OPT_RTAG
,  MY_OPT_RMAP_
,  MY_OPT_TAG
,  MY_OPT_CTAG
,                 MY_OPT_RMAP
,  MY_OPT_SKW =   MY_OPT_RMAP + 2
,  MY_OPT_SKW_
,  MY_OPT_PRM
,  MY_OPT_PRM_
,                       MY_OPT_CHECK
,  MY_OPT_RAWTRANSFORM =  MY_OPT_CHECK + 2
,  MY_OPT_PRMTRANSFORM
,  MY_OPT_SYMTRANSFORM
,  MY_OPT_RV
,  MY_OPT_RE
,  MY_OPT_RI
,  MY_OPT_R
,  MY_OPT_QRE
,  MY_OPT_QRV
,                 MY_OPT_QR
,  MY_OPT_HELP =  MY_OPT_QR + 2
,  MY_OPT_APROPOS
,  MY_OPT_VERSION
}  ;


mcxOptAnchor options[] =
{
   {  "-h"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_HELP
   ,  NULL
   ,  "print this help"
   }
,  {  "--version"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_VERSION
   ,  NULL
   ,  "print version information"
   }
,  {  "--help"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_APROPOS
   ,  NULL
   ,  "print this help"
   }
,  {  "-b"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_BASE
   ,  "<base>"
   ,  "use base.raw, base.hdr, and optionally base.map"
   }
,  {  "-hdr"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_HDR
   ,  "<fname>"
   ,  "read header file"
   }
,  {  "-i"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_SINGLE
   ,  "<fname>"
   ,  "read from single data file"
   }
,  {  "-raw"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RAW
   ,  "<fname>"
   ,  "read raw data file"
   }
,  {  "-map"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MAP
   ,  "<fname>"
   ,  "apply row/col map in file"
   }
,  {  "--map"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_MAP_
   ,  NULL
   ,  "apply row/col map in base.map"
   }
,  {  "-cmap"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CMAP
   ,  "<fname>"
   ,  "apply col map in file"
   }
,  {  "--cmap"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_CMAP_
   ,  NULL
   ,  "apply col map in base.cmap"
   }
,  {  "-rmap"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RMAP
   ,  "<fname>"
   ,  "apply row map in file"
   }
,  {  "--rmap"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_RMAP_
   ,  NULL
   ,  "apply row map in base.cmap"
   }
,  {  "-tag"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_TAG
   ,  "<tag>"
   ,  "apply row/col map in base.tag"
   }
,  {  "-ctag"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CTAG
   ,  "<tag>"
   ,  "apply col map in base.tag"
   }
,  {  "-rtag"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RTAG
   ,  "<tag>"
   ,  "apply row map in base.tag"
   }
,  {  "-skw"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_SKW
   ,  "<fname>"
   ,  "write skew matrix to file"
   }
,  {  "--skw"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_SKW_
   ,  NULL
   ,  "write skew matrix to base.skw"
   }
,  {  "-prm"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_PRM
   ,  "<fname>"
   ,  "write primary result to file"
   }
,  {  "--prm"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_PRM_
   ,  NULL
   ,  "write primary result matrix to base.prm"
   }
,  {  "-raw-tf"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RAWTRANSFORM
   ,  "<func(arg)[, func(arg)]*>"
   ,  "apply unary transformations to raw values"
   }
,  {  "-prm-tf"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_PRMTRANSFORM
   ,  "<func(arg)[, func(arg)]*>"
   ,  "apply unary transformations to primary matrix"
   }
,  {  "-sym-tf"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_SYMTRANSFORM
   ,  "<func(arg)[, func(arg)]*>"
   ,  "apply unary transformations to symmetrified matrix"
   }
,  {  "-digits"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_DIGITS
   ,  "<int>"
   ,  "precision to use in interchange format"
   }
,  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUTPUT
   ,  "<fname>"
   ,  "write to file fname"
   }
,  {  "-xo"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_XO
   ,  "<suf>"
   ,  "write to base.suf (default .sym)"
   }
,  {  "-n"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_NO
   ,  NULL
   ,  "do not write default symmetrized result"
   }
,  {  "+"
   ,  MCX_OPT_HIDDEN
   ,  MY_OPT_PLUS
   ,  NULL
   ,  "write result matrices in binary format"
   }
,  {  "--write-binary"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_WB
   ,  NULL
   ,  "write result matrices in binary format"
   }
,  {  "-s"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_CHECK
   ,  NULL
   ,  "check primary result symmetry by creating skew matrix"
   }
,  {  "-rv"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RV
   ,  "<add|max|mul|left|right>"
   ,  "action for repeated vectors"
   }
,  {  "-re"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RE
   ,  "<add|max|mul|left|right>"
   ,  "action for repeated entries"
   }
,  {  "-ri"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RI
   ,  "<add|max|mul>"
   ,  "action for adding image with mirror"
   }
,  {  "-r"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_R
   ,  "<add|max|mul|left|right>"
   ,  "same for entries and vectors and matrix"
   }
,  {  "--quiet-re"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_QRE
   ,  NULL
   ,  "do not warn for repeated entries"
   }
,  {  "--quiet-rv"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_QRV
   ,  NULL
   ,  "do not warn for repeated vectors"
   }
,  {  "-q"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_QR
   ,  NULL
   ,  "the two above combined"
   }
,  {  NULL, 0, 0, NULL, NULL }
}  ;


#define MAP_COLS 1
#define MAP_ROWS 2
#define MAP_DOMS 4


int main
(  int                  argc
,  const char*          argv[]
)
   {  mcxTing* obase = mcxTingNew("out"), *ibase = NULL
   ;  mcxIO* xf_hdr = NULL, *xf_raw = NULL
      ,  *xf_prm = NULL, *xf_map = NULL, *xf_skew = NULL, *xf_sym = NULL
      ,  *xf_rmap = NULL, *xf_cmap = NULL
   ;  mclVector* dom_rows = mclvNew(NULL, 0), *dom_cols = mclvNew(NULL, 0)
   ;  const char* suf = NULL, *re = NULL, *rv = NULL, *ra = NULL, *ri = NULL
   ;  const char* tag = NULL, *rtag = NULL, *ctag = NULL
   ;  mclx *mx = NULL, *tp = NULL, *sym = NULL, *skew = NULL
   ;  mclx *cmap = NULL, *rmap = NULL
   ;  mcxbits maptype = 0
   ;  mcxbool write_skw =  FALSE
   ;  mcxbool put_sym   =  TRUE
   ;  mcxbool write_prm =  FALSE
   ;  mcxbool check_sym =  FALSE
   ;  mcxbool single_data_file = FALSE
   ;  mcxbits warn_repeat = MCLV_WARN_REPEAT
   ;  int EODATA = EOF
   ;  int digits = MCLXIO_VALUE_GETENV

   ;  void (*ivpmerge)(void* ivp1, const void* ivp2)  =  mclpMergeMax
   ;  double (*fltvecbinary)(pval val1, pval val2)    =  fltMax
   ;  double (*fltmxbinary) (pval val1, pval val2)    =  fltMax

   ;  mclgTF* rawtransform = NULL
   ;  mcxTing* rawtransform_spec = NULL

   ;  mclgTF* symtransform = NULL
   ;  mcxTing* symtransform_spec = NULL

   ;  mclgTF* prmtransform = NULL
   ;  mcxTing* prmtransform_spec = NULL

   ;  mcxstatus parseStatus = STATUS_OK
   ;  mcxOption* opts, *opt

   ;  char* qmode = mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_YES)

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_IO | MCX_LOG_GAUGE | MCX_LOG_WARN
   ;  mclx_app_init(stderr)

   ;  mcxOptAnchorSortById(options, sizeof(options)/sizeof(mcxOptAnchor) -1)

   ;  if
      (!(opts = mcxOptParse(options, (char**) argv, argc, 1, 0, &parseStatus)))
      exit(0)

   ;  for (opt=opts;opt->anch;opt++)
      {  mcxOptAnchor* anch = opt->anch

      ;  switch(anch->id)
         {  case MY_OPT_HELP
         :  case MY_OPT_APROPOS
         :  mcxOptApropos(stdout, me, syntax, 20, MCX_OPT_DISPLAY_SKIP, options)
         ;  return 0
         ;

            case MY_OPT_VERSION
         :  app_report_version(me)
         ;  return 0
         ;

            case MY_OPT_MAP
         :  xf_map = mcxIOnew(opt->val, "r")
         ;  maptype |= MAP_DOMS
         ;  break
         ;

            case MY_OPT_MAP_
         :  maptype |= MAP_DOMS
         ;  break
         ;

            case MY_OPT_PRMTRANSFORM
         :  prmtransform_spec = mcxTingNew(opt->val)
         ;  break
         ;

            case MY_OPT_SYMTRANSFORM
         :  symtransform_spec = mcxTingNew(opt->val)
         ;  break
         ;

            case MY_OPT_RAWTRANSFORM
         :  rawtransform_spec = mcxTingNew(opt->val)
         ;  break
         ;

            case MY_OPT_RMAP
         :  xf_rmap = mcxIOnew(opt->val, "r")
         ;  maptype |= MAP_ROWS
         ;  break
         ;

            case MY_OPT_RMAP_
         :  maptype |= MAP_ROWS
         ;  break
         ;

            case MY_OPT_CMAP
         :  xf_cmap = mcxIOnew(opt->val, "r")
         ;  maptype |= MAP_COLS
         ;  break
         ;

            case MY_OPT_CMAP_
         :  maptype |= MAP_COLS
         ;  break
         ;



            case MY_OPT_SKW
         :  xf_skew = mcxIOnew(opt->val, "w")
         ;  write_skw = TRUE
         ;  break
         ;

            case MY_OPT_SKW_
         :  write_skw = TRUE
         ;  break
         ;

            case MY_OPT_PRM
         :  xf_prm = mcxIOnew(opt->val, "w")
         ;  write_prm = TRUE
         ;  break
         ;

            case MY_OPT_WB
         :  case MY_OPT_PLUS
         :  mclxSetBinaryIO()
         ;  break
         ;

            case MY_OPT_PRM_
         :  write_prm = TRUE
         ;  break
         ;


            case MY_OPT_TAG
         :  tag = opt->val
         ;  break
         ;

            case MY_OPT_CTAG
         :  ctag = opt->val
         ;  break
         ;

            case MY_OPT_RTAG
         :  rtag = opt->val
         ;  break
         ;

            case MY_OPT_SINGLE
         :  xf_hdr = mcxIOnew(opt->val, "r")
         ;  xf_raw = xf_hdr
         ;  single_data_file = TRUE
         ;  EODATA = ')'
         ;  break
         ;

            case MY_OPT_OUTPUT
         :  xf_sym = mcxIOnew(opt->val, "w")
         ;  break
         ;

            case MY_OPT_XO
         :  suf = opt->val
         ;  break
         ;

            case MY_OPT_NO
         :  put_sym = FALSE
         ;  break
         ;

            case MY_OPT_RE
         :  re = opt->val
         ;  break
         ;

            case MY_OPT_RI
         :  ri = opt->val
         ;  break
         ;

            case MY_OPT_RV
         :  rv = opt->val
         ;  break
         ;

            case MY_OPT_R
         :  ra = opt->val
         ;  break
         ;


            case MY_OPT_QR
         :  warn_repeat = 0
         ;  break
         ;

            case MY_OPT_QRE
         :  warn_repeat |= MCLV_WARN_REPEAT_ENTRIES
         ;  warn_repeat ^= MCLV_WARN_REPEAT_ENTRIES
         ;  break
         ;

            case MY_OPT_QRV
         :  warn_repeat |= MCLV_WARN_REPEAT_VECTORS
         ;  warn_repeat ^= MCLV_WARN_REPEAT_VECTORS
         ;  break
         ;


            case MY_OPT_BASE
         :  ibase = mcxTingNew(opt->val)
         ;  mcxTingWrite(obase, opt->val)
         ;  break
         ;

            case MY_OPT_CHECK
         :  check_sym = TRUE
         ;  break
         ;

            case MY_OPT_RAW
         :  xf_raw = mcxIOnew(opt->val, "r")
         ;  break
         ;

            case MY_OPT_HDR
         :  xf_hdr = mcxIOnew(opt->val, "r")
         ;  break
      ;  }
      }

      mcxOptFree(&opts)

   ;  if (single_data_file && xf_raw != xf_hdr)
         mcxErr(me, "some other option conflicts -i usage")
      ,  mcxExit(1)

   ;  if
      (  rawtransform_spec
      && !(rawtransform = mclgTFparse(NULL, rawtransform_spec))
      )
      mcxDie(1, me, "input tf-spec does not parse")

   ;  if
      (  symtransform_spec
      && !(symtransform = mclgTFparse(NULL, symtransform_spec))
      )
      mcxDie(1, me, "sym tf-spec does not parse")

   ;  if
      (  prmtransform_spec
      && !(prmtransform = mclgTFparse(NULL, prmtransform_spec))
      )
      mcxDie(1, me, "prm tf-spec does not parse")

   ;  if (ibase)
      {  if (!xf_raw)
            xf_raw = mcxIOnew(ibase->str, "r")
         ,  mcxIOappendName(xf_raw, ".raw")
      ;  if (!xf_hdr)
            xf_hdr = mcxIOnew(ibase->str, "r")
         ,  mcxIOappendName(xf_hdr, ".hdr")

      ;  if (!xf_map && (tag || maptype & MAP_DOMS))
         {  xf_map = mcxIOnew(ibase->str, "r")
         ;  if (tag)
               mcxIOappendName(xf_map, ".")
            ,  mcxIOappendName(xf_map, tag)
         ;  else
            mcxIOappendName(xf_map, ".map")
      ;  }

      ;  if (!xf_map && !xf_rmap && (rmap || maptype & MAP_ROWS))
         {  xf_rmap = mcxIOnew(ibase->str, "r")
         ;  if (rtag)
               mcxIOappendName(xf_rmap, ".")
            ,  mcxIOappendName(xf_rmap, rtag)
         ;  else
            mcxIOappendName(xf_rmap, ".rmap")
      ;  }

      ;  if (!xf_map && !xf_cmap && (cmap || maptype & MAP_COLS))
         {  xf_cmap = mcxIOnew(ibase->str, "r")
         ;  if (ctag)
               mcxIOappendName(xf_cmap, ".")
            ,  mcxIOappendName(xf_cmap, ctag)
         ;  else
            mcxIOappendName(xf_cmap, ".cmap")
      ;  }
      }

      if (obase)
      {  if (!xf_prm && write_prm)
            xf_prm = mcxIOnew(obase->str, "w")
         ,  mcxIOappendName(xf_prm, ".prm")
      ;  if (!xf_skew && write_skw)
            xf_skew = mcxIOnew(obase->str, "w")
         ,  mcxIOappendName(xf_skew, ".skw")
      ;  if (!xf_sym && put_sym)
            xf_sym = mcxIOnew(obase->str, "w")
         ,  mcxIOappendName(xf_sym, ".")
         ,  mcxIOappendName(xf_sym, suf ? suf : "sym")
   ;  }

      if (ra || re)
      {  re = ra ? ra : re
      ;  if (!strcmp(re, "add"))
         ivpmerge = mclpMergeAdd
      ;  else if (!strcmp(re, "max"))
         ivpmerge = mclpMergeMax
      ;  else if (!strcmp(re, "min"))
         ivpmerge = mclpMergeMin
      ;  else if (!strcmp(re, "mul"))
         ivpmerge = mclpMergeMul
      ;  else if (!strcmp(re, "left"))
         ivpmerge = mclpMergeLeft
      ;  else if (!strcmp(re, "right"))
         ivpmerge = mclpMergeRight
   ;  }

      if (ra || rv)
      {  rv = ra ? ra : rv
      ;  if (!strcmp(rv, "add"))
         fltvecbinary = fltAdd
      ;  else if (!strcmp(rv, "max"))
         fltvecbinary = fltMax
      ;  else if (!strcmp(rv, "min"))
         fltvecbinary = fltMin
      ;  else if (!strcmp(rv, "mul"))
         fltvecbinary = fltCross
      ;  else if (!strcmp(rv, "left"))
         fltvecbinary = fltLeft
      ;  else if (!strcmp(rv, "right"))
         fltvecbinary = fltRight
   ;  }

      if (ra || ri)
      {  ri = ra ? ra : ri
      ;  if (!strcmp(ri, "add"))
         fltmxbinary = fltAdd
      ;  else if (!strcmp(ri, "max"))
         fltmxbinary = fltMax
      ;  else if (!strcmp(ri, "min"))
         fltmxbinary = fltMin
      ;  else if (!strcmp(ri, "mul"))
         fltmxbinary = fltCross
      ;  else if (!strcmp(ri, "left"))
         fltmxbinary = fltLeft
      ;  else if (!strcmp(ri, "right"))
         fltmxbinary = fltRight
   ;  }

      if (!xf_raw || !xf_hdr)
         mcxErr(me, "raw file and header file required")
      ,  mcxExit(1)

   ;  mcxIOopen(xf_hdr, EXIT_ON_FAIL)
   ;  if (xf_raw != xf_hdr)
      mcxIOopen(xf_raw, EXIT_ON_FAIL)
   /* else: -i option was used */

   ;  if (xf_prm)
      mcxIOopen(xf_prm, EXIT_ON_FAIL)

   ;  if (xf_map)
      mcxIOopen(xf_map, EXIT_ON_FAIL)
   ;  else
      {  if (xf_rmap)
         mcxIOopen(xf_rmap, EXIT_ON_FAIL)
      ;  if (xf_cmap)
         mcxIOopen(xf_cmap, EXIT_ON_FAIL)
   ;  }

      if (xf_skew)
      mcxIOopen(xf_skew, EXIT_ON_FAIL)

   ;  if (mclxReadDomains(xf_hdr, dom_cols, dom_rows))
         mcxErr(me, "error parsing header file")
      ,  mcxExit(1)

   ;  if (single_data_file)         /* dangersign, hackish */
      mcxIOexpect(xf_raw, "begin", EXIT_ON_FAIL)

   ;  mx = mclxAllocZero(dom_cols, dom_rows)

   ;  if (xf_map)
         cmap = mclxRead(xf_map, EXIT_ON_FAIL)
      ,  rmap = cmap
      ,  mcxIOclose(xf_map)
   ;  else
      {  if (xf_cmap)
         cmap = mclxRead(xf_cmap, EXIT_ON_FAIL)
      ;  if (xf_rmap)
         rmap = mclxRead(xf_rmap, EXIT_ON_FAIL)
   ;  }

      mclxaSubReadRaw
         (  xf_raw
         ,  mx
         ,  dom_cols
         ,  dom_rows
         ,  EXIT_ON_FAIL
         ,  EODATA
         ,  warn_repeat
         ,  rawtransform ? mclgTFgetEdgePar(rawtransform) : NULL
         ,  ivpmerge
         ,  fltvecbinary
         )
      ,  mcxIOclose(xf_raw)

   ;  if (cmap && mclxMapCols(mx, cmap))
      mcxDie(1, me, "could not map columns")
   ;  if (rmap && mclxMapRows(mx, rmap))
      mcxDie(1, me, "could not map rows")

   ;  if (prmtransform)
      mclgTFexec(mx, prmtransform)

   ;  if (xf_prm)
      mclxWrite(mx, xf_prm, digits, EXIT_ON_FAIL)

   ;  if (put_sym)
      {  tp = mclxTranspose(mx)
      ;  sym = mclxBinary(mx, tp, fltmxbinary)
      ;  if (symtransform)
         mclgTFexec(sym, symtransform)
      ;  mclxWrite(sym, xf_sym, digits, EXIT_ON_FAIL)
      ;  mclxFree(&sym)
               /* not using mclxMergeTranspose because mx is still used */
   ;  }

      if (write_skw || check_sym)
      {  double minone = -1.0
      ;  long n
      ;  tp = tp ? tp : mclxTranspose(mx)
      ;  mclxUnary(tp, fltxMul, &minone)
      ;  skew = mclxAdd(mx, tp)        /* could use mclxMergeTranspose(mx, fltSubtract, 1.0) */
      ;  if (write_skw)
         mclxWrite(skew, xf_skew, digits, EXIT_ON_FAIL)
      ;  n = mclxNrofEntries(skew)
      ;  fprintf(stdout, "symmetry check: %ld skew edges\n", (long) (n/2))
   ;  }

      if (qmode)
      mcxFree(qmode)

   ;  mcxTingFree(&ibase)
   ;  mcxTingFree(&obase)

   ;  if (xf_hdr == xf_raw)
      mcxIOfree(&xf_hdr)
   ;  else
         mcxIOfree(&xf_hdr)
      ,  mcxIOfree(&xf_raw)

   ;  mcxIOfree(&xf_prm)
   ;  mcxIOfree(&xf_map)
   ;  mcxIOfree(&xf_rmap)
   ;  mcxIOfree(&xf_cmap)
   ;  mcxIOfree(&xf_skew)
   ;  mcxIOfree(&xf_sym)

   ;  mclxFree(&mx)
   ;  mclxFree(&tp)
   ;  mclxFree(&sym)
   ;  mclxFree(&skew)

   ;  if (cmap == rmap)
      mclxFree(&cmap)
   ;  else
         mclxFree(&rmap)
      ,  mclxFree(&cmap)

   ;  return 0
;  }


